lib/src/rules/use_build_context_synchronously.dart

```diff
@@ -0,0 +1,199 @@
+  }
+}
+
+class _Visitor extends SimpleAstVisitor {
+  static const _nameBuildContext = 'BuildContext';
+  final Uri _uriFramework;
+
+  // todo (pq): replace in favor of flutter_utils.isBuildContext
+  final LintRule rule;
+
+  _Visitor(this.rule)
+      : _uriFramework = Uri.parse('package:flutter/src/widgets/framework.dart');
+
+  bool accessesContext(ArgumentList argumentList) {
+    for (var argument in argumentList.arguments) {
+      var argType = argument.staticType;
+      if (isBuildContext(argType)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  void check(AstNode node) {
+    // Walk back and look for an async gap that is not guarded by a mounted
+    // property check.
+    var child = node;
+    while (child != null && child is! FunctionBody) {
+      var parent = child.parent;
+      if (parent is Block) {
+        var statements = parent.statements;
+        var index = statements.indexOf(child as Statement);
+        for (var i = index - 1; i >= 0; i--) {
+          var s = statements.elementAt(i);
+          if (isMountedCheck(s)) {
+            return;
+          } else if (isAsync(s)) {
+            rule.reportLint(node);
+            return;
+          }
+        }
+      }
+
+      child = child.parent;
+    }
+  }
+
+  bool isAsync(Statement statement) {
+    if (statement is ExpressionStatement) {
+      var expression = statement.expression;
+      if (expression is AwaitExpression) {
+        return true;
+      }
+    } else if (statement is IfStatement) {
+      return isAsync(statement.thenStatement) ||
+          isAsync(statement.elseStatement);
+    } else if (statement is Block) {
+      for (var s in statement.statements) {
+        if (isAsync(s)) {
+          return true;
+        }
+      }
+    } else if (statement is WhileStatement) {
+      return isAsync(statement.body);
+    } else if (statement is ForStatement) {
+      return isAsync(statement.body);
+    } else if (statement is DoStatement) {
+      return isAsync(statement.body);
+    }
+
+    return false;
+  }
+
+  bool isBuildContext(DartType type) {
+    if (type is! InterfaceType) {
+      return false;
+    }
+    var element = type.element;
+    return element != null &&
+        element.name == _nameBuildContext &&
+        element.source.uri == _uriFramework;
+  }
+
+  bool isMountedCheck(Statement statement) {
+    // This is intentionally naive.  Using a simple 'mounted' property check
+    // as a signal plays nicely w/ unanticipated framework classes that provide
+    // their own mounted checks.  The cost of this generality is the possibility
+    // of false negatives.
+    if (statement is IfStatement) {
+      var condition = statement.condition;
+      if (condition is PrefixExpression) {
+        if (condition.operator.type == TokenType.BANG) {
+          var operand = condition.operand;
+          if (operand is SimpleIdentifier) {
+            if (operand.name == 'mounted') {
+              return true;
```

Comment:
This doesn't check for a return statement. The code `if (!mounted) print('Oops');` would also pass this test.

Reply:
Exactly right (and hilarious).  Fixed.